src/mpl2/src/hier_rtlmp.cpp

```diff
@@ -3045,53 +3032,71 @@ void HierRTLMP::calculateMacroTilings(Cluster* cluster)
-  const float ar = root_cluster_->getHeight() / root_cluster_->getWidth();
-  float max_height = std::sqrt(tilings[0].first * tilings[0].second * ar);
-  float max_width = max_height / ar;
-  max_width = ((root_ux - root_lx) - max_width);
-  max_height = ((root_uy - root_ly) - max_height);
+  const Rect root(root_cluster_->getX(),
+                  root_cluster_->getY(),
+                  root_cluster_->getX() + root_cluster_->getWidth(),
+                  root_cluster_->getY() + root_cluster_->getHeight());
 
-  float std_cell_area = 0.0;
-  for (auto& cluster : root_cluster_->getChildren()) {
-    if (cluster->getClusterType() == StdCellCluster) {
-      std_cell_area += cluster->getArea();
-    }
+  // Note that the range can be larger than the respective core dimension.
+  // As SA only sees what is inside its current outline, this is not a problem.
+  if (io_spans[L].second > io_spans[L].first) {
+    macro_blockages_.emplace_back(root.xMin(),
+                                  io_spans[L].first,
+                                  root.xMin() + depth,
+                                  io_spans[L].second);
+  }
+
+  if (io_spans[T].second > io_spans[T].first) {
+    macro_blockages_.emplace_back(io_spans[T].first,
+                                  root.yMax() - depth,
+                                  io_spans[T].second,
+                                  root.yMax());
+  }
+
+  if (io_spans[R].second > io_spans[R].first) {
+    macro_blockages_.emplace_back(root.xMax() - depth,
+                                  io_spans[R].first,
+                                  root.xMax(),
+                                  io_spans[R].second);
   }
 
-  // Check the range of IO spans. Note that we consider the pins that
-  // are in between boundaries of the core and the die.
-  std::map<PinAccess, std::pair<float, float>> pin_ranges;
-  pin_ranges[L] = std::pair<float, float>(floorplan_uy_, floorplan_ly_);
-  pin_ranges[T] = std::pair<float, float>(floorplan_ux_, floorplan_lx_);
-  pin_ranges[R] = std::pair<float, float>(floorplan_uy_, floorplan_ly_);
-  pin_ranges[B] = std::pair<float, float>(floorplan_ux_, floorplan_lx_);
+  if (io_spans[B].second > io_spans[B].first) {
+    macro_blockages_.emplace_back(io_spans[B].first,
+                                  root.yMin(),
+                                  io_spans[B].second,
+                                  root.yMin() + depth);
+  }
+}
+
+// Determine the range of IOs in each boundary of the die.
+HierRTLMP::IOSpans HierRTLMP::computeIOSpans()
+{
+  IOSpans io_spans;
 
-  int floorplan_lx = micronToDbu(floorplan_lx_, dbu_);
-  int floorplan_ux = micronToDbu(floorplan_ux_, dbu_);
-  int floorplan_uy = micronToDbu(floorplan_uy_, dbu_);
+  odb::Rect die = block_->getDieArea();
+
+  // Initialize spans based on the dimensions of the die area.
+  io_spans[L] = std::pair<float, float>(dbuToMicron(die.yMax(), dbu_),
+                                        dbuToMicron(die.yMin(), dbu_));
```

Comment:
You can write this as :
io_spans[L] = {dbuToMicron(die.yMax(), dbu_), dbuToMicron(die.yMin(), dbu_)};